### PR TITLE
Feature/create user

### PR DESCRIPTION
**Pull request by alresarena2021**
add create user

### CODE REVIEW

File: test/user.spec.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+  before(async function () {
+    await server.startServer();
+  });
+
+  after(async function () {
+    await server.stopServer();
+  });
+
+  it('Create Test', async () => {
```
Comment (ralphcasipe1) [inline test/user.spec.ts:16]: @alresarena2021 
What do you think of having a context in this endpoint? for example,
```ts
describe('Create User',  () => {
  it('creates a new user', () => { ... })
  
  // or 
  it('should create a new user', () => { ... })
})
```

File: test/user.spec.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import axios from 'axios';
+import { Base64 } from 'js-base64';
+import * as server from '../src/server';
+import faker from 'faker';
+describe('REST Test', async () => {
```
Comment (ralphcasipe1) [inline test/user.spec.ts:5]: @alresarena2021 
You can make the title of the test to be more intuitive, for example:
```ts
describe('User Endpoints', () => { ... })
```

File: src/controllers/user-controller.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import { Context } from 'koa';
+import User from '../models/user';
+import { v4 as uuidv4 } from 'uuid';
+import * as bcrypt from 'bcrypt';
+import { isBase64 } from 'class-validator';
+
+export default class UserController {
+  public static async createUser(ctx: Context): Promise<void> {
+    const authData = ctx.request.headers.authorization;
+    const body = JSON.parse(ctx.request.rawBody);
```
Comment (ccpacillos) [inline src/controllers/user-controller.ts:10]: I think you can use here `ctx.request.body` which is already parsed to an object. That way, you don't need to do `JSON.parse` anymore.

File: src/controllers/user-controller.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import { Context } from 'koa';
+import User from '../models/user';
+import { v4 as uuidv4 } from 'uuid';
+import * as bcrypt from 'bcrypt';
+import { isBase64 } from 'class-validator';
+
+export default class UserController {
+  public static async createUser(ctx: Context): Promise<void> {
+    const authData = ctx.request.headers.authorization;
```
Comment (djansyle) [inline src/controllers/user-controller.ts:9]: use `ctx.get` on getting the headers.

**Review by djansyle — CHANGES_REQUESTED** [review]

**Review by ccpacillos — COMMENTED** [review]
You also need to add tests to handle duplicate user emails.

File: src/models/user.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { Schema, model } from 'mongoose';
+import { User } from '../interfaces/user';
+
+const userSchema = new Schema({
+  id: { type: String },
+  email: { type: String },
+  name: { type: String },
+  password: { type: String },
```
Comment (ralphcasipe1) [inline src/models/user.ts:8]: @alresarena2021 I think, `password` and `email` here are `required`.
Is it possible to create a user neither an `email` or a `password`?

File: test/user.spec.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  });
+
+  after(async function () {
+    await server.stopServer();
+  });
+
+  it('Create Test', async () => {
+    const email = faker.internet.email();
+    const password = faker.internet.password();
+    const nameGen = faker.name.findName();
```
Comment (ccpacillos) [inline test/user.spec.ts:19]: Use `name` instead of `nameGen` for a more sensible variable naming.

File: src/controllers/user-controller.ts (hunk 3 above)
Comment (alresarena2021) [reply src/controllers/user-controller.ts:10]: thank you sir clint :)

File: src/controllers/user-controller.ts (hunk 4 above)
Comment (alresarena2021) [reply src/controllers/user-controller.ts:9]: thank you janz :)

File: test/user.spec.ts (hunk 6 above)
Comment (alresarena2021) [reply test/user.spec.ts:19]: thank sir Clint :)

File: test/user.spec.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+          name: nameGen,
+        },
+        {
+          headers: {
+            Authorization: `Basic ${Base64.encode(`${email}:${password}`)}`,
+          },
+        }
+      )
+      .then((result) => {
+        console.log(result.data);
```
Comment (ccpacillos) [inline test/user.spec.ts:33]: Remove `console.log()`.

**Review by ccpacillos — CHANGES_REQUESTED** [review]
There are missing assertions in your test.
Example:
- Verify that the response status is correct.
- Verify that the user was added.

**Review by ralphcasipe1 — CHANGES_REQUESTED** [review]

File: src/controllers/user-controller.ts (hunk 3 above)
Comment (alresarena2021) [reply src/controllers/user-controller.ts:10]: fixed

File: src/controllers/user-controller.ts (hunk 4 above)
Comment (alresarena2021) [reply src/controllers/user-controller.ts:9]: fixed

File: test/user.spec.ts (hunk 6 above)
Comment (alresarena2021) [reply test/user.spec.ts:19]: fixed

File: test/user.spec.ts (hunk 7 above)
Comment (alresarena2021) [reply test/user.spec.ts:33]: fixed

File: test/user.spec.ts (hunk 2 above)
Comment (ralphcasipe1) [reply test/user.spec.ts]: RESOLVED

**Comment by ralphcasipe1** [comment]
> There are missing assertions in your test.
> Example:
> 
> * Verify that the response status is correct.
> * Verify that the user was added.

@ccpacillos  I think this is RESOLVED, can you please check the newly [commit](https://github.com/alresarena2021/rest-api-design/pull/1/commits/be3c5a85d4f4f4dad9acd7f5333526499caec596)

cc: @alresarena2021

File: src/controllers/user-controller.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+          name: body.name,
+          email: email,
+          password: passwordHash,
+        });
+
+        const res = await User.exists({ email: email });
+
+        if (res) ctx.throw(new errorHandler.EmailAlreadyExists());
+
+        await userData.save();
```
Comment (ralphcasipe1) [inline src/controllers/user-controller.ts:31]: @alresarena2021 
This is not a requested change but more of a suggestion.
What do you think of checking the existence of the email first? like this:
```ts
        const res = await User.exists({ email: email });

        if (res) ctx.throw(new errorHandler.EmailAlreadyExists());

        const passwordHash = await bcrypt.hash(password, 3);
        
        await User.create({
          id: uuidv4(),
          name: body.name,
          email: email,
          password: passwordHash,
        });
```
Checking the existence first, would not instantiate the user thus, it's more efficient.

You can make use of `Model#create`, in this case you don't need to `instantiate` +  `Model#save`.

File: src/controllers/user-controller.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+        const passwordHash = await bcrypt.hash(password, 3);
+        const userData = new User({
+          id: uuidv4(),
+          name: body.name,
+          email: email,
+          password: passwordHash,
+        });
+
+        const res = await User.exists({ email: email });
```
Comment (ralphcasipe1) [inline src/controllers/user-controller.ts:27]: @alresarena2021 Can you provide a more intuitive name for this instead of `response` or `res`. You can use `userExisted`.

File: src/controllers/user-controller.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+import User from '../models/user';
+import { v4 as uuidv4 } from 'uuid';
+import * as bcrypt from 'bcrypt';
+import { isBase64 } from 'class-validator';
+import * as errorHandler from '../handlers/error-handler';
+
+export default class UserController {
+  public static async createUser(ctx: Context): Promise<void> {
+    const authData = ctx.request.get('Authorization');
+    const body = ctx.request.body;
```
Comment (ralphcasipe1) [inline src/controllers/user-controller.ts:11]: @alresarena2021  you can refactor this to
```ts
const { body } = ctx.request; 
```

File: src/controllers/user-controller.ts [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+      const encodedData = authArray[1];
+      if (encodedData && isBase64(encodedData)) {
+        const userPass = Buffer.from(encodedData, 'base64').toString();
+        const [email, password] = userPass.split(':');
+
+        const passwordHash = await bcrypt.hash(password, 3);
+        const userData = new User({
+          id: uuidv4(),
+          name: body.name,
+          email: email,
```
Comment (ralphcasipe1) [inline src/controllers/user-controller.ts:23]: @alresarena2021  you can refactor this to:
```ts
{
  ....
  email,
}
```

File: test/user.spec.ts [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+  });
+
+  after(async function () {
+    await server.stopServer();
+  });
+
+  it('should create new user', async () => {
+    const email = faker.internet.email();
+    const password = faker.internet.password();
+    const name = faker.name.findName();
```
Comment (ralphcasipe1) [inline test/user.spec.ts:27]: You can add a helper for this:
```ts
// generate-user
export default function generateUser(param: Pick<User, 'email' | 'name' | 'password'> | {} = {}) {
    return {
         email: faker.internet.email(),
         name: faker.name.findName(),
         password: faker.internet.password(),
         ...params // to override the default returns
    };
}
```

**Review by ralphcasipe1 — CHANGES_REQUESTED** [review]

File: src/controllers/user-controller.ts (hunk 8 above)
Comment (alresarena2021) [reply src/controllers/user-controller.ts:31]: thank you sir ralph :)

File: src/controllers/user-controller.ts (hunk 9 above)
Comment (alresarena2021) [reply src/controllers/user-controller.ts:27]: thank you sir ralph :)

File: src/controllers/user-controller.ts (hunk 10 above)
Comment (alresarena2021) [reply src/controllers/user-controller.ts:11]: thank you sir ralph :)

File: test/user.spec.ts (hunk 12 above)
Comment (alresarena2021) [reply test/user.spec.ts:27]: thank you sir ralph :)

File: src/controllers/user-controller.ts (hunk 8 above)
Comment (alresarena2021) [reply src/controllers/user-controller.ts]: fixed

File: src/controllers/user-controller.ts (hunk 9 above)
Comment (alresarena2021) [reply src/controllers/user-controller.ts]: fixed

File: src/controllers/user-controller.ts (hunk 10 above)
Comment (alresarena2021) [reply src/controllers/user-controller.ts]: fixed

File: src/controllers/user-controller.ts (hunk 11 above)
Comment (alresarena2021) [reply src/controllers/user-controller.ts]: fixed

File: test/user.spec.ts (hunk 12 above)
Comment (alresarena2021) [reply test/user.spec.ts]: fixed

**Review by ralphcasipe1 — APPROVED** [review]

File: test/user.spec.ts [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+          },
+        }
+      )
+    ).to.be.eventually.ok;
+
+    expect(expectResponse.status).to.be.eq(200);
+    expect(expectResponse.data.email).to.be.eq(user.email);
+  });
+
+  it('should not create new user', async () => {
```
Comment (ccpacillos) [inline test/user.spec.ts:44]: I suggest you can make the test name clearer, example:
`should not create new user if email exists`

This way, other developers don't have to read the logic of the test code to know why it should not create a new user.

File: test/user.spec.ts [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    const user = generateUser();
+
+    await User.create(user);
+
+    const expectResponse = await expect(
+      axios.post(
+        URL + '/users',
```
Comment (ccpacillos) [inline test/user.spec.ts:51]: Use template literals: `${URL}/users`.

File: test/user.spec.ts (hunk 13 above)
Comment (alresarena2021) [reply test/user.spec.ts:44]: thanks sir Clint :)

File: test/user.spec.ts (hunk 14 above)
Comment (alresarena2021) [reply test/user.spec.ts:51]: thanks sir clint :)

File: test/user.spec.ts (hunk 13 above)
Comment (alresarena2021) [reply test/user.spec.ts]: fixed

File: test/user.spec.ts (hunk 14 above)
Comment (alresarena2021) [reply test/user.spec.ts]: fixed
